src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java

```diff
@@ -330,7 +330,11 @@ public String getProvidedExtensionAsServer() {
 
   @Override
   public IExtension copyInstance() {
-    return new PerMessageDeflateExtension();
+    PerMessageDeflateExtension clone = new PerMessageDeflateExtension();
+    clone.setThreshold(this.getThreshold());
```

Comment:
Please insert:

+    clone.setDeflater(this.getDeflater());
+    clone.setInflater(this.getInflater());

and replace the new Inflater() calls in lines 169 and 177 with inflater.reset(), and the deflater.end() & new Deflater() calls in lines 247/248 with deflater.reset().

Reply:
We are not going to reuse the deflater/inflater for multiple clients

Reply:
Using reset however should be done, I agree

Reply:
> We are not going to reuse the deflater/inflater for multiple clients

Could you elaborate? If the user of this library wants to hand in a different inflater/deflater, why should that not be used?

Specific use case: I want to use BEST_COMPRESSION, not DEFAULT_COMPRESSION, to get the most out of compression (and both my servers and clients are powerful enough to put up that extra effort). That's why I hand in a new Deflater(Deflater.BEST_COMPRESSION, true) instance. I'd like that to work.

Reply:
The problem is that we create a new instance of a Draft (and therefore the provided extension with the custom deflater/inflater) for every new client connecting to the server.
https://github.com/TooTallNate/Java-WebSocket/blob/master/src/main/java/org/java_websocket/WebSocketImpl.java#L272

If we simply use the single instance provided by the user, we will get in trouble for the server implementation

Reply:
> If we simply use the single instance provided by the user, we will get in trouble for the server implementation

Ah, thanks, I see. I was just lucky that I was only using 1 server and client for my WebSocket compression tests.

Indeed, then an API change is needed to set Inflater and Deflater factories. Or, if we think that this flexibility is not needed, a setCompressionLevel(int level); API would do for me. Then this class would always instantiate the Inflater and Deflater objects, always with nowrap=true.

Reply:
Looking more into the documentation, nowrap will always be true in our case

Reply:
> > If we simply use the single instance provided by the user, we will get in trouble for the server implementation
> 
> Ah, thanks, I see. I was just lucky that I was only using 1 server and client for my WebSocket compression tests.
> 
> Indeed, then an API change is needed to set Inflater and Deflater factories. Or, if we think that this flexibility is not needed, a setCompressionLevel(int level); API would do for me. Then this class would always instantiate the Inflater and Deflater objects, always with nowrap=true.

Yeah I think so too. We only need the compression level, nowrap should be true.

@PhilipRoman you got any objections?

Reply:
My bad, there is no getter for the level/strategy.
setCompressionLevel() is the way to go